### ISSUE TITLE
Safari only: collection not found if replayweb JS is in a cross-origin iframe

### ISSUE THREAD

**Issue by rebeccacremona** [webrecorder/replayweb.page#97]
This one takes a bit of explaining, so please bear with me 🙂.

As discussed at length with @ikreymer, we are working on a setup where users visit https://perma.cc/AAAA-AAAA and are served a page with a banner and an iframe. The iframe is pointed at https://replay.perma.cc, a page on a subdomain that includes an embedded playback of a warc using replayweb.page JS. It's working great in Chrome and Firefox, but does not work in Safari. (Example: https://perma.cc/7ESS-U63W?view-mode=client-side)

It is easy to reproduce outside of our setup: the following HTML just loads https://replayweb.page in an iframe. 
```
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Replayweb.page in an iframe</title>
</head>
<body>
    <iframe style="width:100%; height: 100vh" src="https://replayweb.page">
</body>
</html>
```
If loaded in Firefox or Chrome, you can use the app normally, for instance, to request a playback of https://webrecorder-builds.s3.amazonaws.com/warcs/netpreserve-twitter.warc:
![image](https://user-images.githubusercontent.com/11020492/169871457-1acfe48b-6dc7-4aee-8b64-845d36f6a8ad.png)
Observe the messages in the console: an initial 404 for `https://replayweb.page/w/api/c/id-a3dec314b685?all=1`, followed by the message that the service worker is now registered, followed by a message that 147 records were read, followed by nothing else. And, everything works, you can explore the archive and play back resources perfectly.

But, if loaded in Safari, this [fetch request](https://github.com/webrecorder/replayweb.page/blob/main/src/coll.js#L251) 404s again:
![image](https://user-images.githubusercontent.com/11020492/169872314-42dc984b-d25a-442a-bb35-d23845c30115.png)
Observe the second 404 error in the console, after the message that 147 records were read. The response is indeed from the service worker, as opposed to the live web: `{"error": "collection_not_found"}`.
![image](https://user-images.githubusercontent.com/11020492/169876006-35ed1674-28e6-4c63-ad31-3269626b53ec.png)

Something appears to be going wrong with IndexedDB.... and it has something to do with origins.

Both in the minimal reproduction above (HTML loaded into the browser directly with a `file:///` protocol, iframe pointed at https://replayweb.page) and in my more complex setup (HTML served over SSL, iframe pointed at a subdomain), two instances of `collDB` appear in the Safari dev tools:
![image](https://user-images.githubusercontent.com/11020492/169880138-22d8f8ef-e278-4e55-bfcd-f5658cf973e6.png)

In all other browsers, `CollDB` only appears once. If I visit https://replayweb.page in Safari directly and request a playback of https://webrecorder-builds.s3.amazonaws.com/warcs/netpreserve-twitter.warc, `collDB` only appears once. AND.... if I change my setup so that my playback is not at a subdomain, but rather, both top-level page and the iframe are at the same origin.... then even in Safari `CollDB` only appears once, and even in Safari, the second fetch request succeeds
![image](https://user-images.githubusercontent.com/11020492/169881653-712e27e0-f890-4b5d-8dfa-2bd7c3653e19.png)

I am about out of ideas for how I might be able to fix, work around, or debug further.

- I have tried inspecting the file system to see if multiple sqlite DBs have been created: it doesn't look like it... but it's also not clear to me when Safari flushes this data to disk. 

- I have added a log line to [`_init_DB`](https://github.com/webrecorder/wabac.js/blob/c3009ac6c648ca3f1b5e39715ab5742d5e756672/src/loaders.js#L34) and seen it print once and once only. I have created my own canary DBs in that function and observed that they, too, display twice in the Safari dev tools. 

- I have created a canary DB elsewhere in the code, for instance, in this [constructor](https://github.com/webrecorder/wabac.js/blob/c3009ac6c648ca3f1b5e39715ab5742d5e756672/src/swmain.js#L22), and seen that it does NOT display in duplicate in the Safari dev tools.

Any suggestions?

Note: if you experiment with the minimal reproduction above, you'll see an error to the tune of `Blocked a frame with origin "file:///...." from accessing a frame with origin "https://replayweb.page". Protocols, domains, and ports must match.`: while suggestive, that's not actually the problem. The complete repro, where the iframe's src url has the CORS headers `Access-Control-Allow-Origin=*` and `Access-Control-Allow-Methods=*`, does not therefore throw that error, and yet behaves exactly as described.

**Comment by matteocargnelutti** [comment]
Can confirm that:
- Yes, Safari behaves differently than Chrome / Firefox and doesn't grant storage access by default in that context.
- But sadly no, [doing the necessary](https://perma.minio.test:9000) to get storage access didn't fix the issue. `collDB` still gets split / duplicated. 

(PS: 👋 @ikreymer . I'm new at LIL, excited to be working with Webrecorder tech 😄 ).

**Comment by rebeccacremona** [comment]
Yes! As @matteocargnelutti describes: by throwing in a lot of logging in `wabac.js`'s [`loaders.js`](https://github.com/webrecorder/wabac.js/compare/main...rebeccacremona:worker-contexts?expand=1), I was able to determine that the `collDB` that the web worker was populating was _not_ the same `collDB` that was being used by the service worker during replay, in this very particular context of a cross-origin `<iframe>`.

It's easiest to see by comparing Firefox and Safari, because in those browsers, `console.log` prints to different places, depending on whether the call is made from the service worker or from another context. To see service worker logs in Firefox, you visit `about:debugging#/runtime/this-firefox` and inspect the service worker; in Safari, you select `Develop > Service Workers > (the desired worker)`. In Chrome, on the other hand, everything is printed in the regular dev tools console.... which is normally convenient, but in this case, is not as helpful.

If you start with a completely fresh browser, having made sure to completely clean up any relics from previous playback attempts, a successful playback will include two invocations of [`_initDB`](https://github.com/webrecorder/wabac.js/blob/main/src/loaders.js#L34). In Safari and Firefox, you'll see it logged once in the separate service worker consoles and again, at a later timestamp, in the regular dev console. 

In Firefox and Chrome, as expected and as desired, the upgrade handler is only fired once: that makes sense, since `upgrade` only fires when a db is first created. You'll see the log line, once, in either `about:debugging#/runtime/this-firefox` or Chrome's dev console. 

But in Safari, _the upgrade event handler is being triggered by both invocations of [`_initDB`](https://github.com/webrecorder/wabac.js/blob/main/src/loaders.js#L34)_: once in the separate service worker logs, and then again at a later timestamp in its regular dev tool logs. Which is to say: in Safari, each call is talking to a separate DB; in Firefox and Chrome, they are talking to the same one.

Experimentally, [I reverted the changes](https://github.com/webrecorder/replayweb.page/compare/main...rebeccacremona:worker-contexts?expand=1) made in [the commit that introduced a dedicated web worker for indexing](https://github.com/webrecorder/replayweb.page/commit/178d0507eb08e838334a2919e2822692ef275208). And... playbacks immediately worked in Safari. Without use of the web worker, the log line I added to wabac.js, 
```
console.log(`We are posting the message collAdded, name ${name} at url ${res.config.sourceUrl}`)
```
prints to the service worker-specific console, not the regular Safari dev console; `collDB` does not appear twice in the dev tools (indeed, it actually appears ZERO times...... sigh....).

Wild, right?

Under discussion: what we want to do... Further investigation by @matteocargnelutti suggests that targeting the problematic versions of WebKit within `sw.js` itself may not be feasible; we are thinking about a temporary, somewhat clunky workaround on our end: do the feature detection in our app, and depending on the results, set `replayWebBase` either to the location of a vanilla compilation of replayWebPage, or to the location of a custom compilation that doesn't use a web worker for indexing. We don't love it, but...

Other ideas very welcome!

**Comment by ikreymer** [comment]
Wow, thanks for the detailed investigation.. It seems like whatever it is, likely to do with state partitioning, and perhaps Safari not handling it correctly at all. I was going to suggest doing the test with loading via SW, but you've already done that, nice work!

The loading can be done either via the service worker or a dedicated web worker, I've toggled back and forth between the two: the loading in dedicated worker frees up the service worker to be handling requests, on the other hand, also adds a bit more complexity. At one point, it was removed, but then added back as it was more optimal and addressed some other issues. Currently, both modes are supported though in wabac.js, so it wouldn't be too hard to add that as an option for the replayweb.page to use the service worker-based loading, as you suggest! I agree that would be the best option!
